### ISSUE TITLE
If one out of three AWS instances goes down, then the Minio cluster becomes unresponsive

### ISSUE THREAD

**Issue by davidgeorgealex** [minio/minio#5290]
I have a setup with three EC2 instances, each with 1 docker container with 2 disks.
The cluster starts successfully and I can perform operations on it (using `mc`).
If I `docker kill` a node, then the cluster continue to work, because just 2 out of 6 disks are down.
If I terminate the EC2 instance (from the AWS console UI - stop or terminate), then the cluster becomes unresponsive. I cannot perform any `mc` operations on it (e.q. `mc ls <<alias>>` or `mc admin info <<alias>>). The command just return that the cluster if offline after several minutes. 

Please find bellow how each docker instance is configured:
`
- name: Storage node | Start storage node containers
  docker_container:
    name: "{{ item.1.docker_node_name }}"
    image: "minio/minio:RELEASE.2017-11-22T19-55-46Z"
    detach: true
    ports:
      - "9000:9000"
    restart_policy: always
    network_mode: "host"
    ulimits: nofile:100000:100000
    env:
      MINIO_ACCESS_KEY: "minio"
      MINIO_SECRET_KEY: "minio123"
    command: server http://minio1/export1 http://minio1/export2 http://minio2/export1 http://minio2/export2 http://minio3/export1 http://minio3/export2
  with_indexed_items: "{{ containers_for_host | json_query(filter_query) }}"
  when: item.1.type == 'storage'
  vars:
    filter_query: "[?ip=='{{ inventory_hostname }}'].containers"
    docker_node_name: "{{ item.1.docker_node_name }}"
  become: yes
  become_method: sudo
`

If I start the EC2 instance that was stopped/terminated, then the cluster detects this and becomes back online.

## Expected Behavior
The cluster should be still available, because only 2 out of 6 disks are down.

## Your Environment
* Version used (`minio version`):RELEASE.2017-11-22T19-55-46Z
* MC version: latest
* Environment: EC2

**Comment by vadmeste** [comment]
@davidgeorgealex, does the cluster come back if you wait some minutes without restarting the failed node ?

**Comment by davidgeorgealex** [comment]
@vadmeste It happen like this on the first try, but on the second one it didn't.

**Comment by harshavardhana** [comment]
@davidgeorgealex are you using swarm networking on EC2 instances? also is it possible for you to reach other nodes after the `ec2` instance is down by logging into the nodes which are up? 

Like for example if you `ping` those nodes what is output? - from what i can guess from the behavior EC2 shutdown takes swarm networking down as well and we are hitting some sort of a bug here. 

Just to avoid docker/swarm in the picture is it possible to configure this setup without docker/swarm and just use vanilla binaries instead?

**Comment by davidgeorgealex** [comment]
@harshavardhana I'm not using swarm. If I bring down instance3 for example, from instance1 I'm able to ping instance2 on the host IP. Also from the inside of containers (`docker exec -it <<id>> sh`), I'm able to ping the other containers as well (I got the IPs from `docker network inspect <<network>>`). This is applicable the other way around, pinging from instance2 to instance1. In a matter of latency, I have `64 bytes from 20.1.0.2: seq=0 ttl=64 time=0.465 ms` - pinging docker containers. I will try to see if I can setup without docker.

**Comment by harshavardhana** [comment]
> @harshavardhana I'm not using swarm. If I bring down instance3 for example, from instance1 I'm able to ping instance2 on the host IP. Also from the inside of containers (docker exec -it <<id>> sh), I'm able to ping the other containers as well (I got the IPs from docker network inspect <<network>>). This is applicable the other way around, pinging from instance2 to instance1. In a matter of latency, I have 64 bytes from 20.1.0.2: seq=0 ttl=64 time=0.465 ms - pinging docker containers. I will try to see if I can setup without docker.

Lets say if you try to do `telnet`  ip1 port 9000 and so forth for all ips is it possible? . We have to do this test under two cases as you reproduced

 - Taking ec2 instance down
 - Taking container down using `docker kill`

**Comment by davidgeorgealex** [comment]
@harshavardhana I have 3 IP's 10.0.97.21, 10.0.97.22, 10.0.97.23. 
When all 3 EC2 are up, I get success for all cases (21->22, 21-23, 22->21, ....)
`sudo telnet 10.0.97.23 9000
Trying 10.0.97.21...
Connected to 10.0.97.21.
Escape character is '^]'.`

If I kill container from .23 instance, I get success (as above) between .21 and .22 and to .23 I get:
`sudo telnet 10.0.97.23 9000
Trying 10.0.97.23...
telnet: connect to address 10.0.97.23: Connection refused`

If I stop EC2 instance .23 I get success between .21 and .22 and to .23 I get:
`sudo telnet 10.0.97.23 9000
Trying 10.0.97.23...
telnet: connect to address 10.0.97.23: No route to host`

**Comment by harshavardhana** [comment]
> If I kill container from .23 instance, I get success (as above) between .21 and .22 and to .23 I get:
> sudo telnet 10.0.97.23 9000 Trying 10.0.97.23... telnet: connect to address 10.0.97.23: Connection refused
> 

So the IP is still around just that docker replies back `connection` refused since no registered service.

> If I stop EC2 instance .23 I get success between .21 and .22 and to .23 I get:
> sudo telnet 10.0.97.23 9000 Trying 10.0.97.23... telnet: connect to address 10.0.97.23: No route to host

This is also correct as expected.

**Comment by harshavardhana** [comment]
> @harshavardhana I have 3 IP's 10.0.97.21, 10.0.97.22, 10.0.97.23.
> When all 3 EC2 are up, I get success for all cases (21->22, 21-23, 22->21, ....)
> sudo telnet 10.0.97.23 9000 Trying 10.0.97.21... Connected to 10.0.97.21. Escape character is '^]'.
> 
> If I kill container from .23 instance, I get success (as above) between .21 and .22 and to .23 I get:
> sudo telnet 10.0.97.23 9000 Trying 10.0.97.23... telnet: connect to address 10.0.97.23: Connection refused
> 
> If I stop EC2 instance .23 I get success between .21 and .22 and to .23 I get:
> sudo telnet 10.0.97.23 9000 Trying 10.0.97.23... telnet: connect to address 10.0.97.23: No route to host

Are you doing `telnet` from the host EC2 instance or inside the container?

**Comment by davidgeorgealex** [comment]
from the host.. I will try from the container as well

**Comment by davidgeorgealex** [comment]
Even If all 3 instances are up, from the container, I cannot telnet to another container (the command just wait without any response - `telnet 20.1.0.3 9000`). Also, I cannot telnet to another EC2 instance IP (`telnet 10.0.97.22 9000`). 
The ping is working (`ping 20.1.0.3
PING 20.1.0.3 (20.1.0.3): 56 data bytes
64 bytes from 20.1.0.3: seq=0 ttl=64 time=0.505 ms
64 bytes from 20.1.0.3: seq=1 ttl=64 time=0.514 ms`)

- 10.0.97.21, 10.0.97.22, 10.0.97.23 are the IPs of the EC2s
- 20.1.0.2, 20.1.0.3, 20.1.0.4 are the IPs of the containers

**Comment by davidgeorgealex** [comment]
I tried without Docker (EC2 instances with Minio binaries) and it's working.. so it seams like the issue is related to the Docker communication.

**Comment by harshavardhana** [comment]
> I tried without Docker (EC2 instances with Minio binaries) and it's working.. so it seams like the issue is related to the Docker communication.

Yes that is what i suspected there must be some bug in the network routing when instances are killed off without doing a clean docker shutdown. would it be possible for you to gather docker logs from all instances and share it here. 

We can perhaps try to peek into situations when EC2 is down what happens to other containers communicating in the overlay network.

**Comment by davidgeorgealex** [comment]
I've started all 3 instances, then stop the third one. I've attached the logs. Even if I stopped the third instance, and from the first one I try to run `mc ls`, in the logs I don't see any errors/warnings due to missing disks. I will let this cluster running with one instance down for at least 1h to see what is the result then. Also I will upload again the logs if something will be new there.
[c1.txt](https://github.com/minio/minio/files/1562284/c1.txt)
[c2.txt](https://github.com/minio/minio/files/1562285/c2.txt)
[c3.txt](https://github.com/minio/minio/files/1562286/c3.txt)

**Comment by harshavardhana** [comment]
> I've started all 3 instances, then stop the third one. I've attached the logs. Even if I stopped the third instance, and from the first one I try to run mc ls, in the logs I don't see any errors/warnings due to missing disks. I will let this cluster running with one instance down for at least 1h to see what is the result then. Also I will upload again the logs if something will be new there.
> c1.txt
> c2.txt
> c3.txt

@davidgeorgealex i meant `docker` daemon logs not docker container logs for minio.  You might have to reproduce the hanging issue you mentioned when taking down EC2 instance without using `docker kill`

**Comment by davidgeorgealex** [comment]
@harshavardhana I've reproduced the same "hang" behavior stopping the docker daemon on one instance (`sudo service docker stop`). Now the cluster cannot be used on the other two instances.
I've gather the docker daemon logs before the "stop" and after.
[c1AfterStop.txt](https://github.com/minio/minio/files/1562614/c1AfterStop.txt)
[c1BeforeStop.txt](https://github.com/minio/minio/files/1562615/c1BeforeStop.txt)
[c2AfterStop.txt](https://github.com/minio/minio/files/1562616/c2AfterStop.txt)
[c2BeforeStop.txt](https://github.com/minio/minio/files/1562617/c2BeforeStop.txt)
[c3AfterStop.txt](https://github.com/minio/minio/files/1562618/c3AfterStop.txt)

Next I will try a `docker kill` on the container, which still leads to a working cluster and check the logs again

**Comment by davidgeorgealex** [comment]
Killing and starting the container (`docker kill <<id>>`) doesn't bring anything new to the daemon logs

**Comment by harshavardhana** [comment]
Now trying locally docker-compose even with `docker stop` or `docker kill` things keep running fine @davidgeorgealex 

Looks like from the logs that you shared `serf` tool from consul is interfering with the networking and is sort of managing the network membership, with reconnect attempts every 30 secs and upto 4mins and still progressing.

Here is how i tested locally, start a 4 node setup locally
```
$ docker-compose up
``` 

Start running `minio/mint` functional tests 
```
$ docker run --net=host -e SERVER_ENDPOINT=127.0.0.1:9001 -e ACCESS_KEY=minio -e SECRET_KEY=minio123 -e ENABLE_HTTPS=0 -e MINT_MODE=full minio/mint:edge
Running with
SERVER_ENDPOINT: 127.0.0.1:9001
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       full

To get logs, run 'sudo docker cp f45df4398cbf:/mint/log /tmp/mint-logs'
(1/10) Running aws-sdk-go tests ... done in 0 seconds
(2/10) Running aws-sdk-php tests ... done in 1 minutes and 4 seconds
(3/10) Running aws-sdk-ruby tests ... done in 1 minutes and 29 seconds
(4/10) Running awscli tests ... done in 2 minutes and 14 seconds
(5/10) Running mc tests ... done in 32 seconds
(6/10) Running minio-go tests ... 
```

Now stop one of the containers
```
$  docker stop dockercompose_minio4_1
```

I see that the tests keep running without fail. Some questions regarding docker containers

- Are you using EC2 container service?
- If yes for the first one, Can you try this on a vanilla docker setup i.e EC2 instances with docker installed on them instead? rather than the container service? 

Want to avoid all other tools in this picture.

**Comment by davidgeorgealex** [comment]
@harshavardhana using `docker stop` or `docker kill` on one container, leads to a working setup for me as well. The problem is when I stop EC2 instance (from AWS UI or `shutdown` in the console) or when I stop docker daemon service. Interesting thing with `serf` and consul (we have consul). There is something more I can check on this path?

**Comment by davidgeorgealex** [comment]
Related to your questions, we are not using EC2 container service.

**Comment by davidgeorgealex** [comment]
@harshavardhana Do you think this problem can have the same root cause as https://github.com/minio/minio/issues/5092? The fact that the problem was reproduced in a non-AWS/non-Docker environment made me think yes.

**Comment by harshavardhana** [comment]
Most probably yes.

**Comment by cristic83** [comment]
We retested using the versions: 

minioVersion=RELEASE.2018-01-02T23-07-00Z
mcVersion=RELEASE.2017-12-12T01-08-02Z

and the bug continues to reproduce itself. Please check the attached logs.

[minio-1.log](https://github.com/minio/minio/files/1635125/minio-1.log)
[minio-0.log](https://github.com/minio/minio/files/1635126/minio-0.log)

**Comment by IrinaAndronachi** [comment]
@harshavardhana Are there any updates on this issue?

**Comment by vadmeste** [comment]
@IrinaAndronachi, we are reviewing another code https://github.com/minio/minio/pull/5160 to be merged soon which makes it easier to fix this one.

**Comment by valentinabojan** [comment]
@vadmeste @harshavardhana Hello! Do you have any news about this issue?

**Comment by harshavardhana** [comment]
We can implement this now since a major PR has been merged.

**Comment by valentinabojan** [comment]
@harshavardhana Hi! What's the status of this issue? Are there any chances to be fixed soon? Thank you!

**Comment by harshavardhana** [comment]
> @harshavardhana Hi! What's the status of this issue? Are there any chances to be fixed soon? Thank you!

@valentinabojan there are limitations in Go which do not allow a clean way to implement this right now, we do have a PR which works but it is based on arbitrary timeouts which is basically more trouble than what we want here.

**Comment by StefanVanDyck** [comment]
@harshavardhana Hello, are you guys still working on this?
Your last reply made use wonder if no fix was coming.
If we can assist in any way, let us know.

Thanks!

**Comment by harshavardhana** [comment]
@StefanVanDyck we are we had to refactor the entire RPC sub-system and move to full HTTP to make it work correctly. You can follow the PR here https://github.com/minio/minio/pull/5840

**Comment by harshavardhana** [comment]
The PR is merged and this issue should be fixed as of #5840
